homeassistant/components/shelly/__init__.py

```diff
@@ -83,59 +69,111 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
 
     coap_context = await get_coap_context(hass)
 
+    # First check if device is not sleeping
     try:
         async with async_timeout.timeout(AIOSHELLY_DEVICE_TIMEOUT_SEC):
             device = await aioshelly.Device.create(
                 aiohttp_client.async_get_clientsession(hass),
                 coap_context,
                 options,
+                True,
             )
     except (asyncio.TimeoutError, OSError) as err:
-        raise ConfigEntryNotReady from err
+        # Not a sleeping device, raise error
+        if not entry.data["sleep_period"]:
+            raise ConfigEntryNotReady from err
+
+    # Sleeping device, make an empty device
+    if device is None:
+        async with async_timeout.timeout(AIOSHELLY_DEVICE_TIMEOUT_SEC):
+            device = await aioshelly.Device.create(
+                aiohttp_client.async_get_clientsession(hass),
+                coap_context,
+                options,
+                False,
+            )
 
-    hass.data[DOMAIN][DATA_CONFIG_ENTRY][entry.entry_id] = {}
     coap_wrapper = hass.data[DOMAIN][DATA_CONFIG_ENTRY][entry.entry_id][
         COAP
     ] = ShellyDeviceWrapper(hass, entry, device)
     await coap_wrapper.async_setup()
 
-    hass.data[DOMAIN][DATA_CONFIG_ENTRY][entry.entry_id][
-        REST
-    ] = ShellyDeviceRestWrapper(hass, device)
+    # All ready, perform full setup
+    if device.initialized:
+        _LOGGER.debug("Device ready - IP:%s", ip_address)
+        hass.data[DOMAIN][DATA_CONFIG_ENTRY][entry.entry_id][
+            REST
+        ] = ShellyDeviceRestWrapper(hass, device)
+
+        for component in PLATFORMS:
+            hass.async_create_task(
+                hass.config_entries.async_forward_entry_setup(entry, component)
+            )
+        return True
 
-    for component in PLATFORMS:
+    _LOGGER.debug("Device not ready - IP:%s", ip_address)
+
+    # Setup only sensors for sleeping device
+    for component in SLEEPING_PLATFORMS:
         hass.async_create_task(
             hass.config_entries.async_forward_entry_setup(entry, component)
         )
 
     return True
 
 
+async def async_migrate_entry(hass: HomeAssistant, entry: ConfigEntry):
```

Comment:
We shouldn't do this with a migration. We can just do this in async_setup_entry if we see that the key is not there. But I think that we should just skip migration and only say that battery devices are supported after this PR and let people repair their old ones ?

Reply:
Since the user needs to wake up the device anyhow first time after this PR, I think your approach is better, we should update the docs that sleeping devices which added before HA with this PR should be removed and added again, what do you think?

Reply:
Migration removed, will only add data if it is missing